joomla/joomla-cms
# [fix] store JModelList filters as objects in session without relaying in Registry

This issue tries to fix issue https://github.com/joomla/joomla-cms/issues/6544 after updating Registry package to latest framework version.

Replaces https://github.com/joomla/joomla-cms/pull/6548
## A bit of history

When I implemented searchtools (yes you can blame me :dancer: ) I stored filters in session as array and relayed in Registry to cast them to objects. After updating to the latest Registry package my modification was gone and searchtools started to experience conflicts between filters sent as objects (the old way) and filters sent as arrays (the new way). 

Having reviewed and discussed about it I now see that I did it wrong forcing Registry to cast data to objects. The solution is to cast data in JModelList directly. But can we now do that without causing B/C issues with extensions already using searchtools? This what this pull request tries to test. Let's see if the best fix is still possible. If not we will need a B plan.
## Backward compatibility issues

As you can see in the changes the best behavior `parent::populateState()` is called now on top of any model extending `JModelList` that way automatic filters can be overriden by custom filters. In the old implementation the `parent::populateState` was done the last thing to allow that any filter could initialize the object in session before any filter is stored. That was mainly to support old filtering systems (like the ones that Hathor uses).

After searchtools was implemented extension developers should decide if you they going to use searchtools or not and based in that setup their model `populateState()` method.

The issue may now be that third part extensions have to move the `parent::populateState()` method inside their model `populateState()`. So **we need third part extensions using searchtools to HEAVILY test this**.
## Why is this the best fix possible?

Because it now ensures that all the filters are stored in session in the same way (objects). Filters can be accessed & modified in the same way than before Searchtools was implemented (using objects). The idea is that JModelList automatically detects the filters allowing you to override them if needed. 
## How to test it

Go to any of the current managements using Searchtools and ensure that filters work normally. This includes:
- Creating more than 5 items per management to be able to test pagination
- Set any filter and then test pagination ensuring that filters are kept and pagination works normally.
- Reset filters should restore all the values and should redirect you to page 1 of search results.
- Enable Hathor template and test that all the filters work there too. Thos filters are not using searchtools. Before you had to logout / login to get the filters working but now it should work correctly.

Managements to test:
- com_banners > Banner list
- com_banners > Client list
- com_content > Category list
- com_content > Article list
- com_content > Featured articles list
- com_menus > Item list
## Test by third part extensions
- Install your extension in a joomla that includes this patch and test that your filters work normally.
